src/slurmdbd_ops.py

```diff
@@ -0,0 +1,448 @@
+=cs1s
+-----END PGP PUBLIC KEY BLOCK-----
+"""
+
+
+class Slurmdbd:
+    """Facilitate slurmdbd package lifecycle ops."""
+
+    def __init__(self):
+        self._package_name = "slurmdbd"
+        self._keyring_path = Path("/usr/share/keyrings/slurm-wlm.asc")
+
+    def _repo(self) -> apt.DebianRepository:
+        """Return the slurmctld repo."""
+        ppa_url: str = "https://ppa.launchpadcontent.net/ubuntu-hpc/slurm-wlm-23.02/ubuntu"
+        sources_list: str = (
+            f"deb [signed-by={self._keyring_path}] {ppa_url} {distro.codename()} main"
+        )
+        return apt.DebianRepository.from_repo_line(sources_list)
+
+    def install(self) -> None:
+        """Install the slurmdbd package using lib apt."""
+        # Install the key.
+        if self._keyring_path.exists():
+            self._keyring_path.unlink()
+        self._keyring_path.write_text(SLURM_PPA_KEY)
+
+        # Add the repo.
+        repositories = apt.RepositoryMapping()
+        repositories.add(self._repo())
+
+        # Install the slurmctld, slurm-client packages.
+        try:
+            # Run `apt-get update`
+            apt.update()
+            apt.add_package([self._package_name])
+        except apt.PackageNotFoundError:
+            logger.error(f"{self._package_name} not found in package cache or on system")
+        except apt.PackageError as e:
+            logger.error(f"Could not install {self._package_name}. Reason: %s", e.message)
+
+    def uninstall(self) -> None:
+        """Uninstall the slurmdbd package using libapt."""
+        # Uninstall the slurmdbd package.
+        if apt.remove_package(self._package_name):
+            logger.info(f"{self._package_name} removed from system.")
+        else:
+            logger.error(f"{self._package_name} not found on system")
+
+        # Disable the ubuntu-hpc repo.
+        repositories = apt.RepositoryMapping()
+        repositories.disable(self._repo())
+
+        # Remove the key.
+        if self._keyring_path.exists():
+            self._keyring_path.unlink()
+
+    def upgrade_to_latest(self) -> None:
+        """Upgrade slurmctld to latest."""
+        try:
+            slurmdbd = apt.DebianPackage.from_system(self._package_name)
+            slurmdbd.ensure(apt.PackageState.Latest)
+            logger.info("updated vim to version: %s", slurmdbd.version.number)
+        except apt.PackageNotFoundError:
+            logger.error("a specified package not found in package cache or on system")
+        except apt.PackageError as e:
+            logger.error("could not install package. Reason: %s", e.message)
+
+    def version(self) -> str:
+        """Return the slurmdbd version."""
+        try:
+            slurmdbd_vers = apt.DebianPackage.from_installed_package(self._package_name)
+        except apt.PackageNotFoundError:
+            logger.error(f"{self._package_name} not found on system")
+        return slurmdbd_vers.version.number
+
+
+class SlurmdbdOpsManager:
```

Comment:
So one thing I'm wondering here is if we should condense the API of this class here. Since the operations management is now scoped per charm, it might make sense to scope an operation manager to each service we have within the charm. For example, we could have an operations manager for both `munge` and `slurmdbd`.

Then the method signatures could shorter like the following:

```python
class SlurmdbdManager:
    def install(...): ...

    def remove(...): ...
    
    def update(...): ...
```

What do you think?